### PR TITLE
feat: add dropdown component

### PR DESCRIPTION
**Pull request by NadiaPosch**

### CODE REVIEW

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.tsx
Comment (dwirz) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.tsx:13]: Try to add/use a union of two types for small and large

**Comment by dwirz** [comment]
align styling of trigger for hover and disable

**Review by dwirz — CHANGES_REQUESTED** [review]

File: src/components/Dropdown/Dropdown.tsx
Comment (dwirz) [inline src/components/Dropdown/Dropdown.tsx:35]: from a semantic perspective IMO this should be a `<button>` (you then potentially need to wrap your content within a `<span>` because grid/flex does not work as expected for a button 🙈 )

File: src/components/Dropdown/Dropdown.tsx
Comment (dwirz) [inline src/components/Dropdown/Dropdown.tsx]: There is also another way you could do that but not sure it is more readable
`[css.trigger, !activeItem && css.inactive, size === Size.Large && css.large].filter(Boolean).join(" " )`

File: src/components/Dropdown/DropdownMenu/DropdownMenu.tsx
Comment (dwirz) [inline src/components/Dropdown/DropdownMenu/DropdownMenu.tsx:22]: same as above

File: src/components/Dropdown/Dropdown.spec.tsx
Comment (dwirz) [inline src/components/Dropdown/Dropdown.spec.tsx:122]: can't we just for the prop and ignore the style itself?

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.stories.tsx
Comment (dwirz) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.stories.tsx:20]: for boolean cases you could add
```
args: {
  warning: false,
  disabled: false,
}
```
this way storybook creates a bool toggle

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.stories.tsx
Comment (dwirz) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.stories.tsx:33]: if you add `warning` and `disabled` to `default.args` above you can omit it here

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.stories.tsx
Comment (dwirz) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.stories.tsx:42]: same as above

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.tsx
Comment (dwirz) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.tsx:67]: for completeness i would add `&& !disabled`

File: src/components/Dropdown/Dropdown.tsx
Comment (NadiaPosch) [reply src/components/Dropdown/Dropdown.tsx]: I have extracted it into a function as in the `MenuItem` component. It's more readable that way in my opinion.

File: src/components/Dropdown/DropdownMenu/DropdownMenu.tsx
Comment (NadiaPosch) [reply src/components/Dropdown/DropdownMenu/DropdownMenu.tsx:22]: Since there are only two different classes here I don't think the readablility would be improved by joining it together in an array. Therefore I'd leave it as it is.

File: src/components/Dropdown/Dropdown.spec.tsx
Comment (NadiaPosch) [reply src/components/Dropdown/Dropdown.spec.tsx:122]: I tried but somehow it only works when specifying a style 🤷🏽‍♀️

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.stories.tsx
Comment (dwirz) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.stories.tsx:16]: The `onClick`
```
    args: {
        //...
        onClick: { action: "onClear" },
    },
```
should be an `argTypes`
```
    argTypes: { 
      onClick: { action: "Item Clicked" },
    },
```
otherwise you get the following error `Uncaught TypeError: onClick is not a function`

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.module.css
Comment (dwirz) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.module.css:76]: sonar told me that this prop is duplicated 😏

File: src/components/Dropdown/DropdownMenu/DropdownMenu.tsx
Comment (oliverschwendener) [inline src/components/Dropdown/DropdownMenu/DropdownMenu.tsx:38]: Is there a reason behind using object destructuring here for the `menuItem`? I think if we would just use 

```tsx
{menuItems.map((menuItem) => (
    <DropdownMenuItem
        key={menuItem.id}
        onClick={() => onChange(menuItem.id)}
        title={menuItem.title}
        icon={menuItem.icon}
        subtitle={menuItem.subtitle}
        size={menuItem.size}
        style={menuItem.style}
        disabled={menuItem.disabled}
        active={menuItem.id === activeItemId}
    />
))}
```

it would be easier to read and understand where each variable is coming from. Additionally it would help avoiding naming collisions with other variables that are defined in an upper scope. For example when using object destructuring the variable `size` on line 26 is colliding with the variable `size` from line 18.

@NadiaPosch What do you think? :)

**Review by dwirz — APPROVED** [review]

File: src/components/Dropdown/DropdownMenu/DropdownMenu.tsx
Comment (NadiaPosch) [reply src/components/Dropdown/DropdownMenu/DropdownMenu.tsx:38]: I try to destructure whenever it's possible to reduce redundancy but if you prefer the version above that's fine for me :)

Comment (oliverschwendener) [reply src/components/Dropdown/DropdownMenu/DropdownMenu.tsx]: In this case I'd prefer not to destructure the object because we are otherwise overwriting variables from the upper scope.

**Comment by oliverschwendener** [comment]
@NadiaPosch Something that I just noticed: should the dropdown menu hide when the dropdown loses focus, for example when the user clicks outside of the dropdown?

File: src/components/Dropdown/DropdownMenu/DropdownMenu.tsx
Comment (SamuelAlev) [inline src/components/Dropdown/DropdownMenu/DropdownMenu.tsx:24]: We may want to have an `id` instead of using the index to cache the rendered elements better.

```
Array of MenuItem:
0. [...]
1. [...] -> deleted
2. [...] -> get rerendered even without changes
```

**Review by oliverschwendener — CHANGES_REQUESTED** [review]
See my comment above.

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.tsx
Comment (SamuelAlev) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.tsx:70]: I would add a `user-select: none;` on both element

File: src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.module.css
Comment (SamuelAlev) [inline src/components/Dropdown/DropdownMenuItem/DropdownMenuItem.module.css:51]:
```suggestion
```

I don't think it is needed as we have the `grid-gap` set and I don't think it affects anything

File: src/components/Dropdown/Dropdown.tsx
Comment (SamuelAlev) [inline src/components/Dropdown/Dropdown.tsx:80]: We miss the contextual options part

Capture d’écran 2021-07-15 à 16 45 37

**Review by SamuelAlev — CHANGES_REQUESTED** [review]

File: src/components/Dropdown/DropdownMenu/DropdownMenu.tsx
Comment (dwirz) [reply src/components/Dropdown/DropdownMenu/DropdownMenu.tsx:24]: i told @NadiaPosch to do it like this for the moment, but you're right and we're with you on this, it's better to change the items from `MenuItem[][]` to something like `{id: string; items: MenutItem[] }[]` this it would also allow to potentially add a title per group or other props

File: src/components/Dropdown/Dropdown.tsx
Comment (dwirz) [reply src/components/Dropdown/Dropdown.tsx:80]: since atm we do not actually know what should happen if someone click on the dots i would add them later in a seperate PR wdyt?

File: src/components/Dropdown/DropdownMenu/DropdownMenu.tsx
Comment (NadiaPosch) [reply src/components/Dropdown/DropdownMenu/DropdownMenu.tsx]: Done 👌🏼

File: src/components/Dropdown/Dropdown.tsx
Comment (dwirz) [inline src/components/Dropdown/Dropdown.tsx:14]: Move this to `DropdownMenu` otherwise you have a circular dependency between `Dropdown` and `DropdownMenu`. Also I would name it `MenuBlock`.

Comment (NadiaPosch) [reply src/components/Dropdown/Dropdown.tsx:14]: Done 👌🏼

File: src/components/Dropdown/Dropdown.tsx
Comment (NadiaPosch) [reply src/components/Dropdown/Dropdown.tsx:94]: I agree with Dominique and have also talked to Simone about it. Maybe we can make a note somewhere to make sure it won't be overlooked when the design is ready.

**Review by oliverschwendener — APPROVED** [review]
I'll be off for a week starting today. I'll approve this PR so you don't have to wait for my approval. 😉

File: src/components/Dropdown/Dropdown.stories.tsx
Comment (oliverschwendener) [inline src/components/Dropdown/Dropdown.stories.tsx:35]: Shouldn't this be `menuBlocks`?

File: src/components/Dropdown/Dropdown.stories.tsx
Comment (oliverschwendener) [inline src/components/Dropdown/Dropdown.stories.tsx:98]: And also this?

File: src/components/Dropdown/Dropdown.stories.tsx
Comment (NadiaPosch) [reply src/components/Dropdown/Dropdown.stories.tsx]: Yes, I just fixed it 🙈

File: src/components/Dropdown/Dropdown.spec.tsx
Comment (SamuelAlev) [inline src/components/Dropdown/Dropdown.spec.tsx:89]:
```suggestion
    placeholder,
    initialActiveId,
```
It can be `undefined`

File: src/components/Dropdown/Dropdown.module.css
Comment (SamuelAlev) [inline src/components/Dropdown/Dropdown.module.css:11]:
```suggestion
.triggerWrapper {
```

**Review by SamuelAlev — CHANGES_REQUESTED** [review]
We should have `12px` instead of `20px` for the padding of the text (large and small) and right caret-down have `12px` instead of `8px` (small version).
Capture d’écran 2021-07-16 à 14 39 09

The `grid-gap`/`gap` should be `12px`
Capture d’écran 2021-07-16 à 14 51 10


[I also asked to Simone about MenuItem disabled vs hovering colors](https://www.figma.com/file/v00soRe37Tdb14Nno9VFEl?node-id=6086:190565#94717216) as they are the same and makes it weird for the eyes.

**Comment by NadiaPosch** [comment]
@SamuelAlev Where did you find the spacing examples exactly? In the Figma file that is linked in the ticket, I measure `20px` for small and large items 🤔
https://www.figma.com/file/v00soRe37Tdb14Nno9VFEl/Arcade?node-id=6086%3A190565

**Comment by NadiaPosch** [comment]
I just saw that there are two types of selection indicator icons, so I'll quickly add it as an option
image

**Comment by SamuelAlev** [comment]
@NadiaPosch https://www.figma.com/file/v00soRe37Tdb14Nno9VFEl/Arcade?node-id=4588%3A211854

Seems the Dropdown has 12px but the MenuItem has 16px, problably a quick sync with Simone is worth.

**Pull request merged**
